File: app/src/main/scala/tripaint/view/TriPaintView.scala
package tripaint.view

import tripaint.Color
import tripaint.coords.StorageCoords
import tripaint.grid.GridCell
import tripaint.grid.ImageSaveCollisionHandler
import tripaint.image.format.StorageFormat

import java.io.File

trait TriPaintView extends ImageSaveCollisionHandler {
  def backgroundColor: Color

  def askForImageSize(): Option[Int]
  def askSaveBeforeClosing(images: Seq[GridCell]): Option[Boolean]
  def askForWhereToPutImage(): Option[(Int, Int)]

  def askForSaveFile(image: GridCell): Option[File]
  def askForFileSaveSettings(file: File, image: GridCell): Option[FileSaveSettings]

  def askForFileToOpen(): Option[File]
  def askForFileOpenSettings(
      file: File,
      imageSize: Int,
      xCount: Int,
      yCount: Int
  ): Option[FileOpenSettings]

  def askForBlurRadius(): Option[Int]
  def askForMotionBlurRadius(): Option[Int]
  def askForRandomNoiseColors(): Option[(Color, Color)]

  def close(): Unit
}

case class FileSaveSettings(offset: StorageCoords, format: StorageFormat)
case class FileOpenSettings(offset: StorageCoords, format: StorageFormat)

File: ui/src/main/kotlin/tripaint/view/TriPaintView.kt
package tripaint.view

import tripaint.color.Color
import tripaint.coords.StorageCoords
import tripaint.grid.GridCell
import tripaint.grid.ImageSaveCollisionHandler
import tripaint.image.format.StorageFormat
import java.io.File

interface TriPaintView : ImageSaveCollisionHandler {
    fun backgroundColor(): Color

    fun askForImageSize(): Int?
    fun askSaveBeforeClosing(images: List<GridCell>): Boolean?
    fun askForWhereToPutImage(): Pair<Int, Int>?

    fun askForSaveFile(image: GridCell): File?
    fun askForFileSaveSettings(file: File, image: GridCell): FileSaveSettings?

    fun askForFileToOpen(): File?
    fun askForFileOpenSettings(
        file: File,
        imageSize: Int,
        xCount: Int,
        yCount: Int
    ): FileOpenSettings?

    fun askForBlurRadius(): Int?
    fun askForMotionBlurRadius(): Int?
    fun askForRandomNoiseColors(): Pair<Color, Color>?

    fun close()
}

data class FileSaveSettings(val offset: StorageCoords, val format: StorageFormat)
data class FileOpenSettings(val offset: StorageCoords, val format: StorageFormat)
